proposed/activity-improvements.md

```diff
@@ -0,0 +1,393 @@
+        public bool Equals(ActivityContext context);
+        public override bool Equals(object? obj);
+        public override int GetHashCode();
+    }
+}
+```
+
+### ActivityEvent
+```C#
+
+namespace System.Diagnostics
+{
+    /// A text annotation associated with a collection of attributes.
+    /// Event has name and optionally has timestamps and attributes
+    public sealed class ActivityEvent
+    {
+        /// Initializes a new instance of the <see cref="ActivityEvent"/> class.
+        public ActivityEvent(string name) : this(name, DateTimeOffset.UtcNow, s_emptyAttributes);
+        public ActivityEvent(string name, DateTimeOffset timestamp) : this(name, timestamp, s_emptyAttributes);
+        public ActivityEvent(string name, IReadOnlyDictionary<string, object> attributes);
+        public ActivityEvent(string name, DateTimeOffset timestamp, IReadOnlyDictionary<string, object> attributes);
+
+        public string Name { get; }
+        public DateTimeOffset Timestamp { get; }
+        public IReadOnlyDictionary<string, object> Attributes { get; }
+    }
+}
+```
+
+ ### ActivityLink
+
+```C#
+namespace System.Diagnostics
+{
+    /// Activity may be linked to zero or more other <see cref="ActivityContext"/> that are causally related.
+    /// Links can point to ActivityContexts inside a single Trace or across different Traces.
+    /// Links can be used to represent batched operations where a Activity was initiated by multiple initiating Activities,
+    /// each representing a single incoming item being processed in the batch.
+    public readonly struct ActivityLink
+    {
+        public ActivityLink(ActivityContext context);
+        public ActivityLink(ActivityContext context, IDictionary<string, object>? attributes);
+        public ActivityContext Context { get; }
+        public IDictionary<string, object>? Attributes { get; }
+    }
+}
+```
+
+### ActivityCreationOptions
+
+```C#
+namespace System.Diagnostics
+{
+    // ActivityCreationOptions contains values telling if decide to create the Activity object
+    // and with what state
+    // This enum value will be returned from the callback of the Activity listener object which created by ActivitySource.AddActivityListener.
+    public enum ActivityCreationOptions
+    {
+        /// Don't allow creating the Activity object.
+        NoCreate,
+
+        /// Create the Activity but drop recording events and tags.
+        NoDataAllowed,
+
+        /// Create the Activity and allow recording events and tags
+        AllowData,
+
+        /// Create the Activity, allow recording events and tags, and turn on ActivityTraceFlags.Recorded (i.e. Allow Sampling).
+        AllowDataAndSampling
+    }
+}
+```
+
+### ActivitySource
+
+```C#
+
+namespace System.Diagnostics
+{
+    /// ActivitySource is the type that help create and start Activity and allow listening to the Activity events too.
+    public sealed class ActivitySource : IDisposable
```

Comment:
@pakrym requested that we have ActivitySource.HasListeners() (or some similar name for the same functionality) that offers a super fast inlinable check if the Source has any listeners. If this function returned false it can save the instrumented code from needing to spend the time to create all the arguments to the ActivitySource.StartActivity() function. In particular it saves the time to read the ActivityContext from its wire format and convert it into an id or ActivityContext which takes some parsing and allocation.

Reply:
good catch, I missed this one. I'll add it

Reply:
done